locator/ec2_snitch.cc

```diff
@@ -66,11 +68,57 @@ future<> ec2_snitch::start() {
 future<sstring> ec2_snitch::aws_api_call(sstring addr, uint16_t port, sstring cmd) {
     return connect(socket_address(inet_address{addr}, port))
     .then([this, addr, cmd] (connected_socket fd) {
+        _sd = std::move(fd);
+        _in = _sd.input();
+        _out = _sd.output();
+        _token_req = sstring("PUT ") + cmd +
+                     sstring(" HTTP/1.1\r\nHost: ") + addr +
+                     sstring("\r\nX-aws-ec2-metadata-token-ttl-seconds: 21600") +
+                     sstring("\r\n\r\n");
+        return _out.write(_token_req.c_str()).then([this] {
+             return _out.flush();
+        });
+     }).then([this] {
+        _parser.init();
+        return _in.consume(_parser).then([this] {
+            if (_parser.eof()) {
+                return make_exception_future<sstring>("Bad HTTP response");
+            }
+
+            // Read HTTP response header first
+            auto _rsp = _parser.get_parsed_response();
+            auto rc = _rsp->_status_code;
+            // Verify EC2 instance metadata access
+            if (rc == 403) {
+                throw std::runtime_error("Error: Unauthorized response received when trying to communicate with instance metadata service.");
+            }
+
+            auto it = _rsp->_headers.find("Content-Length");
+            if (it == _rsp->_headers.end()) {
+                return make_exception_future<sstring>("Error: HTTP response does not contain: Content-Length\n");
+            }
+
+            auto content_len = std::stoi(it->second);
+
+            // Read HTTP response body
+            return _in.read_exactly(content_len).then([this] (temporary_buffer<char> buf) {
+                sstring res(buf.get(), buf.size());
+
+                return make_ready_future<sstring>(std::move(res));
+            });
+        });
+    });
+}
```

Comment:
This copying is unwarranted. Can we merge it back into the original aws_api_call? Perhaps with a `std::optional<sstring> token = {}` parameter.

Reply:
Agreed -- it makes sense. I've been continuation'ized while writing this. I'll simplify.